### PR TITLE
httproute: correct conformance for regex support

### LINKED ISSUE #190: Regex support in HTTP path and header matching
Regular expression support for route matching is a common feature request.

A little bit of history:

Ingress v1beta1 had regex support for posix regex and also had strict(er) path validation.
This was seen as a limitation because some implementation didn't support posix regex but instead supported PCRE regex.

We need to clarify follow in our API:
1. Should regex support be in the "core", "extended" or "custom" part of the API?
2. [Currently](https://github.com/kubernetes-sigs/service-apis/blob/96e09e85cc0ced3fb6f41dbd16d885a42970e11b/api/v1alpha1/httproute_types.go#L90-L91), we have a `RegularExpression` and `ImplementationSpecific` as two different path types. Does it make sense to fold Regex into Implementation specific? Or should there be `PCRERegex` and `POSIXRegex` as two different path types, there by providing a little more portability between implementations?

### PR DESCRIPTION
**Pull request by hbagdi**
Changes:
- Move regex support from extend to custom. The details in regex make it
  hard to share the support in extended or core.
- add enum based validation for match type, this can be relaxed in
future to allow for arbitrary strings
- introduce kubebuilder annotations

### CODE REVIEW

File: apis/v1alpha1/generated.proto [hunk 1]
```diff
@@ -602,12 +602,13 @@ message HTTPRouteMatch {
...
   // Support: core (Exact, Prefix)
-  // Support: extended (RegularExpression)
-  // Support: custom (ImplementationSpecific)
+  // Support: custom (RegularExpression, ImplementationSpecific)
   //
   // Default: "Exact"
   //
   // +optional
+  // +kubebuilder:validation:Enum=Exact;Prefix;RegularExpression;ImplementationSpecific
+  // +kubebuilder:default=Exact
```
Comment (robscott) [inline apis/v1alpha1/generated.proto:611]: Do these work in the correct order? I know other things like required annotations do not work with default values.

**Comment by robscott** [comment]
@hbagdi Thanks for filing this! This mostly LGTM, just had a question about CRD validation and defaulting ordering. One other note, this is the first time I've seen the `do-not-merge/invalid-commit-message` label, maybe it doesn't like the bulleted list?

File: apis/v1alpha1/generated.proto (hunk 1 above)
Comment (jpeach) [reply apis/v1alpha1/generated.proto:611]: I hope this works, we use it in `HostnameMatch` and `GatewayAddress` too.

File: apis/v1alpha1/httproute_types.go [hunk 2]
```diff
@@ -139,6 +140,8 @@ type HTTPRouteMatch struct {
 	// Default: "Exact"
 	//
 	// +optional
+	// +kubebuilder:validation:Enum=Exact;ImplementationSpecific
+	// +kubebuilder:default=Exact
```
Comment (jpeach) [inline apis/v1alpha1/httproute_types.go:144]: Adding a Go type can make this clearer for API users:

```Go
// Pretty sure that the validation is transitive here ...
// +kubebuilder:validation:Enum=Exact;ImplementationSpecific
type HeaderMatchType string

// We already have these constants, just need to add types.
const (
ImplementationSpecificHeaderMatch HeaderMatchType = "ImplementationSpecific"
ExactHeaderMatch HeaderMatchType = "Exact"
)

type HTTPRouteMatch struct {
...
// +kubebuilder:default=Exact
HeaderMatchType *string `json:"headerMatchType" protobuf:"bytes,3,opt,name=headerMatchType"`
}

```

File: apis/v1alpha1/generated.proto (hunk 1 above)
Comment (hbagdi) [reply apis/v1alpha1/generated.proto:611]: I applied the CRD and the basic example and it correctly populated the default. I'm using k8s 1.16.9.

File: apis/v1alpha1/httproute_types.go (hunk 2 above)
Comment (hbagdi) [reply apis/v1alpha1/httproute_types.go:144]: Can we do this in a separate PR?

**Comment by hbagdi** [comment]
This should be good for another round of review.

**Comment by robscott** [comment]
Thanks!

/lgtm

**Comment by hbagdi** [comment]
Rebased once again. Please review.

**Comment by hbagdi** [comment]
Sorry, clicked the wrong button by mistake.

**Pull request merged**